### ISSUE TITLE
Short round-trips have the outgoing trip skipped

### ISSUE THREAD

**Issue by shankari** [e-mission/e-mission-server#302]
Kids took my phone, went to the library and came back. Return trip showed up, outgoing trip did not.
Return trip was 12:53 to 12:57.

trip coming back: `57798b0488f66332d48de971`

When we entered the CLEAN_RESAMPLING stage, we had:

```
2016-07-03 22:00:35,252:DEBUG:Returning entry with length 3 result
2016-07-03 22:00:35,253:DEBUG:Considering place 57798b0288f66332d48de961: 2016-07-03T12:12:08.010000-07:00 -> 2016-07-03T12:17:06.872000-07:00
2016-07-03 22:00:35,253:DEBUG:Considering place 57798b0288f66332d48de963: 2016-07-03T12:19:59.140000-07:00 -> 2016-07-03T12:53:17.999000-07:00
2016-07-03 22:00:35,253:DEBUG:Considering place 57798b0288f66332d48de965: 2016-07-03T12:57:55-07:00 -> None

2016-07-03 22:00:35,253:DEBUG:Considering trip 57798b0288f66332d48de960: 2016-07-03T12:10:46.617000-07:00 -> 2016-07-03T12:12:08.010000-07:00
2016-07-03 22:00:35,254:DEBUG:Considering trip 57798b0288f66332d48de964: 2016-07-03T12:53:17.999000-07:00 -> 2016-07-03T12:57:55-07:00
```

So the earlier trips look like they were stripped out as part of the cleaning stage. Why is this?

Started with the the trip leaving at 12:10:46 from [-122.0838776, 37.3904947].
This is already the library.

The matching place before this is:

```
2016-07-03 22:00:35,271:DEBUG:retrieved object Entry({u'data': {u'enter_fmt_time': u'2016-07-01T19:45:33.126000-07:00', u'exit_fmt_time': u'2016-07-03T12:10:46.617000-07:00', u'ending_trip': ObjectId('57773c6c88f663773188f683'), u'starting_trip': ObjectId('57798b0288f66332d48de960'), u'location': {u'type': u'Point', u'coordinates': [-122.086378, 37.3907813]}, u'enter_ts': 1467427533.126, u'duration': 145513.49100017548, u'exit_ts': 1467573046.617}, u'_id': ObjectId('57773c6c88f663773188f684'), u'metadata': {u'write_fmt_time': u'2016-07-01T21:00:44.551724-07:00', u'write_ts': 1467432044.551724, u'time_zone': u'America/Los_Angeles', u'platform': u'server', u'key': u'segmentation/raw_place'}}) and added to id_map
```

So the raw_place was itself broken, the error is not in the cleaning stage.

Here's why the raw_place is broken. The main reason is that the differences between points is too small, probably because people were walking slowly or waiting at the signal or whatever.

```
2016-07-03 22:00:34,154:INFO:Points
AttrDict({u'loc': {u'type': u'Point', u'coordinates': [-122.0860081, 37.3918007]},
u'fmt_time': u'2016-07-03T12:02:49.571000-07:00',
'metadata_write_ts': 1467572573.296,
u'ts': 1467572569.5710001,
'_id': ObjectId('57797daafe930a83ace95b58'),
u'accuracy': 20.533999999999999})
AttrDict({u'loc': {u'type': u'Point', u'coordinates': [-122.08576, 37.3916086]},
u'fmt_time': u'2016-07-03T12:03:11-07:00',
'metadata_write_ts': 1467572590.678,
u'ts': 1467572591.0,
'_id': ObjectId('57797daafe930a83ace95b5a'),
u'accuracy': 10.0})
are within the distance filter and only 1 min apart so part of the same trip
```
But the first point should have been outside the geofence filter because of the value of the geofence radius....

However, it turns out that in this case, we compared the first point to a point that occurred after it!

```
2016-07-03 22:00:34,151:DEBUG:------------------------------2016-07-03T12:02:49.571000-07:00------------------------------
2016-07-03 22:00:34,151:DEBUG:Appending currPoint because the current start point is None
2016-07-03 22:00:34,152:DEBUG:Comparing with prev_point = AttrDict({u'loc': {u'type': u'Point', u'coordinates': [-122.0864586, 37.3910005]},
u'fmt_time': u'2016-07-03T13:00:53-07:00',
'metadata_write_ts': 1467576051.553,
'_id': ObjectId('57797daafe930a83ace95c22')})
```

why is the prev_point for `2016-07-03T12:02:49.571000-07:00` set to `2016-07-03T13:00:53-07:00`?

`2016-07-03T13:00:53-07:00` looks like it is the last point read in this set...

```
2016-07-03 22:00:33,915:DEBUG:check_prior_duplicate called with size = 73, entry = 78
2016-07-03 22:00:33,919:INFO:Found existing filtered location for entry at index = 78, id = 57797daafe930a83ace95c21, ts = 1467576053.0, fmt_time = 2016-07-03T13:00:53-07:00, skipping
2016-07-03 22:00:33,921:INFO:For stage PipelineStages.ACCURACY_FILTERING, last_ts_processed = 2016-07-03T20:00:55.908000
```

I see why, and it's actually a pretty serious and annoying bug. 

Here's the code for comparing against the prev point. Note that for the first point, `just_ended = True`, and `idx = 0`, so we will end up comparing the first point with the last point from that batch. As long as the first and last points are sufficiently far apart, we are fine. But if the first and last points overlap, as they would for a short round trip, then we will skip the first leg of the trip.

```
        just_ended = True
        for idx, row in filtered_points_df.iterrows():
            currPoint = ad.AttrDict(row)
            currPoint.update({"idx": idx})
            logging.debug("-" * 30 + str(currPoint.fmt_time) + "-" * 30)
            if curr_trip_start_point is None:
                logging.debug("Appending currPoint because the current start point is None")
                # segmentation_points.append(currPoint)

            if just_ended:
                prev_point = ad.AttrDict(filtered_points_df.iloc[idx - 1])

```

**Comment by shankari** [comment]
Some potential fixes:

1. use the location of the last place as the initial prev point
1. ignore the just_ended check if idx == 0

However, we need to be careful that we don't regress the case in which we have a spurious trip and start getting all kinds of weird trips. Will have to test the changes carefully.

An example of a spurious trip was this one at around 2pm.

Again, we compared with the last point and detected that the trip continued.

```
2016-06-21 22:00:28,318:INFO:**********UUID 0763de67-f61e-3f5d-90e7-518e69793954: segmenting into trips**********
2016-06-21 22:00:28,554:DEBUG:Comparing with prev_point = AttrDict({u'loc': {u'type': u'Point',
 u'coordinates': [-122.0863361, 37.3910994]},
 u'fmt_time': u'2016-06-21T14:08:35.465000-07:00',
 'metadata_write_ts': 1466543316.2509999,
 u'altitude': 0.0,
 u'ts': 1466543315.4649999,
 u'longitude': -122.0863361,
 u'filter': u'time',
 u'latitude': 37.391099400000002,
 '_id': ObjectId('5769b2114a9bba50e3fb4110'),
 u'heading': 0,
 u'sensed_speed': 0.0,
 u'accuracy': 90.019999999999996})
```

After several comparisons, though, we found a point that was further way and
started a trip. 

```
2016-06-21 22:00:28,570:DEBUG:------------------------------2016-06-21T14:05:17.080000-07:00------------------------------
2016-06-21 22:00:28,572:DEBUG:Setting new trip start point AttrDict({u'loc': {u'type': u'Point', u'coordinates': [-122.086499, 37.3911307]},
 'idx': 11,
 u'fmt_time': u'2016-06-21T14:05:17.080000-07:00',
 'metadata_write_ts': 1466543117.884,
 u'altitude': -7.899993896484375,
 u'ts': 1466543117.08,
 u'longitude': -122.086499,
 u'filter': u'time',
 u'latitude': 37.3911307,
 '_id': ObjectId('5769b2114a9bba50e3fb40f5'),
 u'heading': 318,
 u'sensed_speed': 1.06335,
 u'accuracy': 5.0}) with idx 11
```

The trip ended fairly soon after that.
2016-06-21 22:00:28,625:INFO:Found trip end at 2016-06-21T14:06:02.846000-07:00

We created a raw trip

```
2016-06-21 22:00:28,721:DEBUG:Inserting entry Entry({'data': Rawtrip({'end_place': ObjectId('5769b8fc88f6636d6c59fa5c'),
 'start_ts': 1466543117.08,
 'start_place': ObjectId('5768a05d88f66324c58b1915'),
 'duration': 45.766000032424927,
 'start_loc': {"coordinates": [-122.086499, 37.3911307], "type": "Point"},
 'end_ts': 1466543162.846,
 'start_fmt_time': u'2016-06-21T14:05:17.080000-07:00',
 'end_loc': {"coordinates": [-122.0863238, 37.3908003], "type": "Point"},
 'source': 'DwellSegmentationTimeFilter',
 'end_fmt_time': u'2016-06-21T14:06:02.846000-07:00',
 '_id': ObjectId('5769b8fc88f6636d6c59fa5b'),
 'user_id': UUID('0763de67-f61e-3f5d-90e7-518e69793954'),
 'metadata': Metadata({'write_fmt_time': '2016-06-21T15:00:28.718152-07:00',
 'write_ts': 1466546428.718152,
 'time_zone': 'America/Los_Angeles',
 'platform': 'server',
 'key': 'segmentation/raw_trip'})}) into timeseries
```

But it had only one activity and it was STILL, so no unfiltered activities.

```
2016-06-21 22:00:28,764:INFO:++++++++++++++++++++Processing trip 5769b8fc88f6636d6c59fa5b for user 0763de67-f61e-3f5d-90e7-518e69793954++++++++++++++++++++
2016-06-21 22:00:28,812:DEBUG:Found 1 results
2016-06-21 22:00:28,816:DEBUG:curr activity = Motionactivity({u'confidence': 100, u'fmt_time': u'2016-06-21T14:05:41.211000-07:00', u'ts': 1466543141.211, 'metadata_write_ts': 1466543141.211, u'local_dt': {u'hour': 14, u'month': 6, u'second': 41, u'weekday': 1, u'year': 2016, u'timezone': u'America/Los_Angeles', u'day': 21, u'minute': 5}, '_id': ObjectId('5769b2114a9bba50e3fb40f8'), u'type': 3})
2016-06-21 22:00:28,816:DEBUG:filtered points []
```

**Comment by shankari** [comment]
As an example of how careful we need to be about the change, consider another missing trip - the trip back from Bella's house after dropping off at the birthday party. I originally thought that the problem was the same, but it turns out that it isn't.

The trip to her house ended at 

```
2016-06-22 00:00:21,378:INFO:Found trip end at 2016-06-21T15:54:37.544000-07:00
```

Trip coming back was at:

```
2016-06-22 00:00:21,381:DEBUG:Setting new trip start point...u'2016-06-21T16:00:40.251000-07:00'
2016-06-22 00:00:21,917:INFO:Found trip end at 2016-06-21T16:06:20.105000-07:00
```

We create a raw trip for this:

```
2016-06-22 00:00:22,029:DEBUG:Inserting entry Entry({'data': Rawtrip({'end_place': ObjectId('5769d51688f6637d8e8eebe3'),
 'start_ts': 1466550040.251,
 'start_place': ObjectId('5769d51688f6637d8e8eebe1'),
 'duration': 339.85400009155273,
 'start_loc': {"coordinates": [-122.0768071, 37.3882945], "type": "Point"},
 'end_ts': 1466550380.105,
 'start_fmt_time': u'2016-06-21T16:00:40.251000-07:00',
 'end_loc': {"coordinates": [-122.0861172, 37.3907863], "type": "Point"},
 'end_fmt_time': u'2016-06-21T16:06:20.105000-07:00',
 '_id': ObjectId('5769d51688f6637d8e8eebe2'),
 'metadata': Metadata({'write_fmt_time': '2016-06-21T17:00:22.026304-07:00',
 'write_ts': 1466553622.026304,
 'time_zone': 'America/Los_Angeles',
 'platform': 'server',
 'key': 'segmentation/raw_trip'})}) into timeseries
```

There are no sections, so we filter it out in cleaning. Why are there no sections?


There are 11 MotionActivity results

```
2016-06-22 00:00:22,272:INFO:++++++++++++++++++++Processing trip 5769d51688f6637d8e8eebe2 for user 0763de67-f61e-3f5d-90e7-518e69793954++++++++++++++++++++
2016-06-22 00:00:22,316:DEBUG:curr_query = {'$or': [{'metadata.key': 'background/motion_activity'}], 'user_id': UUID('0763de67-f61e-3f5d-90e7-518e69793954'), 'data.ts': {'$lte': 1466550380.105, '$gte': 1466550040.251}}, sort_key = data.ts
2016-06-22 00:00:22,317:DEBUG:Found 11 results
```

But after filtering there is only one point, so the section start and end are
the same, and there are no points for the section, so we skip it.

```
2016-06-22 00:00:22,324:DEBUG:filtered points [1]
2016-06-22 00:00:22,341:DEBUG:Considering MotionTypes.BICYCLING from 2016-06-21T16:01:01.946000-07:00 -> 2016-06-21T16:01:01.946000-07:00
2016-06-22 00:00:22,343:WARNING:Found no location points between
Motionactivity({u'confidence': 69,
 u'fmt_time': u'2016-06-21T16:01:01.946000-07:00',
 u'ts': 1466550061.9460001,
 'metadata_write_ts': 1466550061.9460001,
 '_id': ObjectId('5769d0ae4a9bba50e3fb4bb7'),
 u'type': 1})
and 
Motionactivity({u'confidence': 69,
 u'fmt_time': u'2016-06-21T16:01:01.946000-07:00',
 u'ts': 1466550061.946,
 'metadata_write_ts': 1466550061.946,
 '_id': ObjectId('5769d0ae4a9bba50e3fb4bb7'),
 u'type': 1})
```

**Comment by shankari** [comment]
The challenge is to:
- drop the trip at 14:00
- keep the trip at 16:00
- keep the trip at noon on the 3rd

**Comment by shankari** [comment]
Found new missing trip but it seems fairly long - the Berkeley to Millbrae leg on the 15th of July is missing.

**Comment by shankari** [comment]
Ran into this again with the trip to the optometrist, even when we ran a repro run, because the last point loaded was again home, so it was effectively a VERY long round trip.


```
2016-08-14 14:21:44,896:DEBUG:------------------------------2016-08-10T09:53:59.425000-07:00------------------------------
2016-08-14 14:21:44,897:DEBUG:Comparing with prev_point = AttrDict({u'local_dt_hour': 19, u'local_dt_timezone': u'America/Los_Angeles', u'local_dt_second': 22, u'local_dt_year': 2016, u'local_dt_month': 8, u'local_dt_weekday': 2, u'elapsedRealtimeNanos': 435581721000000, u'loc': {u'type': u'Point', u'coordinates': [-122.0863187, 37.390994]}, 'user_id': UUID('0763de67-f61e-3f5d-90e7-518e69793954'), u'altitude': 0.0, u'ts': 1470883882.0610001, u'latitude': 37.390993999999999, u'sensed_speed': 0.0, u'accuracy': 19.408999999999999, u'fmt_time': u'2016-08-10T19:51:22.061000-07:00', u'local_dt_minute': 51, 'metadata_write_ts': 1470883883.2939999, u'longitude': -122.08631870000001, u'filter': u'time', '_id': ObjectId('57abf481f311e474bfda0918'), u'heading': 0, u'local_dt_day': 10})
```

When this ran in the real pipeline, the last point would have been one at 11, but the principle is the same.

```
2016-08-14 14:21:44,898:INFO:Points AttrDict({'fmt_time': u'2016-08-10T19:51:22.061000-07:00'}) and AttrDict({'fmt_time': u'2016-08-10T09:53:59.425000-07:00'}) are within the distance filter and only 1 min apart so part of the same trip
```

And then that continues until the return trip

```
2016-08-14 14:21:44,935:DEBUG:------------------------------2016-08-10T11:07:05.684000-07:00------------------------------
2016-08-14 14:21:44,936:DEBUG:Appending currPoint because the current start point is None
2016-08-14 14:21:44,936:DEBUG:Comparing with prev_point = AttrDict({u'local_dt_hour': 10, u'local_dt_timezone': u'America/Los_Angeles', u'local_dt_second': 53, u'local_dt_year': 2016, u'local_dt_month': 8, u'local_dt_weekday': 2, u'elapsedRealtimeNanos': 400329834833126, u'loc': {u'type': u'Point', u'coordinates': [-122.0799955, 37.3915915]}, 'user_id': UUID('0763de67-f61e-3f5d-90e7-518e69793954'), u'altitude': -12.399993896484375, u'ts': 1470848633.0, u'latitude': 37.391591499999997, u'sensed_speed': 3.4402499999999998, u'accuracy': 10.0, u'fmt_time': u'2016-08-10T10:03:53-07:00', u'local_dt_minute': 3, 'metadata_write_ts': 1470848643.02, u'longitude': -122.0799955, u'filter': u'time', '_id': ObjectId('57ab63cef311e474bfd97270'), u'heading': 332, u'local_dt_day': 10})
2016-08-14 14:21:44,937:DEBUG:Setting new trip start point AttrDict({u'local_dt_hour': 11, u'local_dt_timezone': u'America/Los_Angeles', u'local_dt_second': 5, u'local_dt_year': 2016, u'local_dt_month': 8, u'local_dt_weekday': 2, u'elapsedRealtimeNanos': 404125792000000, u'loc': {u'type': u'Point', u'coordinates': [-122.0817472, 37.3918142]}, 'user_id': UUID('0763de67-f61e-3f5d-90e7-518e69793954'), u'altitude': 0.0, u'ts': 1470852425.684, u'latitude': 37.3918142, u'sensed_speed': 0.0, u'accuracy': 20.046, u'fmt_time': u'2016-08-10T11:07:05.684000-07:00', u'local_dt_minute': 7, 'metadata_write_ts': 1470852435.133, 'idx': 23, u'longitude': -122.0817472, u'filter': u'time', '_id': ObjectId('57ab75f3f311e474bfd97dd9'), u'heading': 0, u'local_dt_day': 10}) with idx 23
```

**Comment by shankari** [comment]
From the potential fixes listed above:

1. use the location of the last place as the initial prev point
2. ignore the just_ended check if idx == 0

I am going to go with (2). It feels more principled because:

1. the check for continuation was added to deal with trailing points after a trip end. idx = 0, not to avoid starting spurious trips.
1. we have other mechanisms to deal with spurious trips - e.g. checking whether there are sections
1. Even in the 2pm spurious trip example above, the glomming on only worked for some time - we eventually did start a (spurious) trip that was cleaned by later steps.

At least I'm going to try it for a bunch of days in July and August and see how
well it works.

> After several comparisons, though, we found a point that was further way and
started a trip.
> We created a raw trip
> But it had only one activity and it was STILL, so no unfiltered activities.

**Comment by shankari** [comment]
Re-running prior data, to see if we can fix the prior bugs.
Recall that our goal was:
- drop the trip at 14:00
- keep the trip at 16:00
- keep the trip at noon on the 3rd

Here are the results for that. As we can see, we do drop the trip at 14:00 and keep the trip at 16:00 (at least, the trip that ends at 16:00) but we combine them into one trip, probably because of #352. We do keep the trip at noon on the 3rd. Note however that there's something weird going on, and the trips are not ordered correctly any more.

| Jun 21 | July 3 |
| -------| ------ |
| ![simulator screen shot aug 16 2016 12 40 05 pm](https://cloud.githubusercontent.com/assets/2423263/17713382/7888e64a-63af-11e6-8e42-793e75266f19.png) | ![simulator screen shot aug 16 2016 12 42 55 pm](https://cloud.githubusercontent.com/assets/2423263/17713383/788a764a-63af-11e6-8aba-c68166f8380c.png) |

**Comment by shankari** [comment]
#355 fixes #354, but still does not have the correct order.
We have the trip from our house to PYT, continuing on to Bella's house and arriving at 3:54.
We should then have a trip from Bella's house to the library at 4. Instead, that trip shows up at 5:13.

The expectation is that trip coming back was at 

```
2016-06-22 00:00:21,381:DEBUG:Setting new trip start point...u'2016-06-21T16:00:40.251000-07:00'
2016-06-22 00:00:21,917:INFO:Found trip end at 2016-06-21T16:06:20.105000-07:00
```
